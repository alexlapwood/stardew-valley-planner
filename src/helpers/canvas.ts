import { getSeason } from "./date";

export function getCanvasImages(date: number) {
  const season = ["spring", "summer", "fall", "winter"][getSeason(date)];

  const createImage: HTMLImageElement | null = document.querySelector(
    `img[src="/images/create.png"]`
  );

  const destroyImage: HTMLImageElement | null = document.querySelector(
    `img[src="/images/destroy.png"]`
  );
  const backgroundImage: HTMLImageElement | null = document.querySelector(
    `img[src="/images/background-${season}.png"]`
  );
  const cropsImage: HTMLImageElement | null = document.querySelector(
    `img[src="/images/crops.png"]`
  );

  if (
    backgroundImage === null ||
    createImage === null ||
    cropsImage === null ||
    destroyImage === null
  ) {
    throw new Error("Error loading images");
  }

  return { backgroundImage, createImage, cropsImage, destroyImage };
}

export function getCanvasPositionAndScale(canvas?: HTMLCanvasElement) {
  if (canvas === undefined) {
    throw new Error();
  }

  const rect = canvas.getBoundingClientRect();
  const { left, top } = rect;

  const scaleX = canvas.width / (rect.right - rect.left);
  const scaleY = canvas.height / (rect.bottom - rect.top);

  return { left, top, scaleX, scaleY };
}
