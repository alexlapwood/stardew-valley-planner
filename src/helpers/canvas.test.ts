import { getCanvasImages, getCanvasPositionAndScale } from "./canvas";

describe("getCanvasImages", () => {
  it("throws an error if there are no images", () => {
    expect(getCanvasImages).toThrowErrorMatchingSnapshot();
  });
});

describe("getCanvasPositionAndScale", () => {
  document.body.innerHTML = "<canvas />";
  const canvas = document.querySelector("canvas");

  if (canvas === null) {
    throw new Error();
  }

  it("", () => {
    const result = getCanvasPositionAndScale(canvas);
    expect(result).toEqual({
      left: 0,
      scaleX: Infinity,
      scaleY: Infinity,
      top: 0
    });
  });
});
