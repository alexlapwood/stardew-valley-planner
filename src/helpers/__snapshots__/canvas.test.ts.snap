// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getCanvasImages throws an error if there are no images 1`] = `"Error loading images"`;
